Instrumentation via graphql-ruby's tracer feature

bmorton opened the pull request:
Building off the implementation in #19 for the feature discussed in #18, this moves the tracing into the storage adapters so that they can be accurately tagged with the adapter name that triggered the event.

Also different from #19 is how some events (cache hit and cache miss) are traced with empty blocks so that they can simply emit things that happened, but don't have any code to wrap.  I'd love feedback on this.  It works well (we're using this experimental branch in production right now, though clients haven't fully adopted APQ yet) but it feels a bit awkward.  Like you mentioned before, this is an advanced feature so we may be able to deal with some awkwardness here.

This is also disabled behind a configuration flag, so by default it's essentially a no-op with essentially zero performance overhead.

I'm going to add a couple questions in code to clarify some things, but I think this is overall a better approach.  Let me know what you think and I'm happy to write up some docs (probably outside of the README for this?) and address any feedback.

bmorton commented on lib/graphql/persisted_queries/schema_patch.rb:
```diff
@@ -41,7 +44,12 @@ def verify_http_method=(verify)
+      end
+
       def multiplex(queries, **kwargs)
+        persisted_query_store.tracers = tracers if persisted_queries_tracing_enabled?
```
This is a little awkward, but something like this is necessary since we don't know if tracers will be configured before or after our plugin is configured (it is dependent on how someone sets up their schema).  An alternative would be to patch `GraphQL::Schema.tracer` to pass through to also set the tracer on the store.

bmorton replied:
I ended up moving this to a patch on `GraphQL::Schema.tracer` and it's a bit clearer now.

bmorton commented on lib/graphql/persisted_queries/store_adapters/memcached_store_adapter.rb:
```diff
@@ -17,11 +17,23 @@ def initialize(dalli_client:, expiration: nil, namespace: nil)
+                trace("fetch_query.cache_hit", adapter: :memcached)
+              else
+                trace("fetch_query.cache_miss", adapter: :memcached)
+              end
```
Instead of doing something like this in every `StoreAdapter`, we could move this to the `BaseStoreAdapter` with some minor modifications and then we could avoid the duplication.

Something like this:

```ruby
class BaseStoreAdapter
  # ...

  def fetch_query(_hash)
    fetch(_hash).tap do |result|
      if result
        trace("fetch_query.cache_hit", adapter: @name)
      else
        trace("fetch_query.cache_miss", adapter: @name)
      end
    end
  end

  # ...

  private

  def fetch(_hash)
    raise NotImplementedError
  end
end
```

And then have `fetch` and `save` (or something similar) defined in each of the adapters.  That'd clean up some of the boilerplate.

DmitryTsepelev replied:
Agreed, great idea!

bmorton replied:
This is implemented now as well!

bmorton commented on spec/graphql/persisted_queries/schema_patch_spec.rb:
```diff
@@ -14,6 +14,24 @@ def call(error)
     end
   end
 
+  Tracer = Class.new do
```
Currently. the only tests that I added for this are these integration-y tests.  It felt awkward testing it in each of the adapters, so I haven't added any there.  From the previous comment, if we move most of the implementation to the `BaseStorageAdapter`, it would make more sense to add a few tests there.  What do you think?

DmitryTsepelev replied:
We could probably move tests to the shared context and include for both adapter, but overall I agree that tracing should be tested separetely

bmorton replied:
Since the implementation moved to `BaseStoreAdapter`, I've added the tests there and they seem like they are in the right place to me.

bmorton commented:
> Yeah, I was planning to split the Readme into separate docs, my plan is to create the docs folder, put the tracing doc there and link it from the readme (any place), and I'll migrate the rest sections as soon as this PR is merged. What do you think?

Sounds good!  I'll create `docs/tracing.md` and link to that from the README.

bmorton commented:
This should now be good to go!  I've fixed the things discussed above, added documentation, and updated the CHANGELOG.  Let me know if there's anything else!

DmitryTsepelev commented on lib/graphql/persisted_queries/store_adapters/base_store_adapter.rb:
```diff
@@ -5,15 +5,45 @@ module PersistedQueries
+
+        def fetch_query(hash)
+          fetch(hash).tap do |result|
+            if result
```
```suggestion
             event = result ? "cache_hit" : "cache_miss"
             trace("fetch_query.#{event}", adapter: @name)
```

bmorton replied:
I'm good with it.  Done!

bmorton commented:
Argh, you're right, forgot to push.  Everything should be pushed and good now!

DmitryTsepelev commented:
Will make a release soon, thank you so much!
